core/src/com/unciv/logic/civilization/CivilizationInfo.kt

```diff
@@ -713,6 +742,22 @@ class CivilizationInfo {
         addNotification("[${otherCiv.civName}] gave us a [${militaryUnit.name}] as gift near [${city.name}]!", locations, otherCiv.civName, militaryUnit.name)
     }
 
+    /** Gain a random great person from a random city state */
+    private fun gainGreatPersonFromCityState() {
+        val givingCityState = getKnownCivs().filter { it.isCityState() && it.getAllyCiv() == civName}.random()
+        val giftedUnit = gameInfo.ruleSet.units.values.filter { it.isGreatPerson() }.random()
+        val cities = NextTurnAutomation.getClosestCities(this, givingCityState)
+        val placedUnit = placeUnitNearTile(cities.city1.location, giftedUnit.name)
+        if (placedUnit == null) return
+        val locations = LocationAction(listOf(placedUnit.getTile().position, cities.city2.location))
+        addNotification( "[${givingCityState.civName}] gave us a [${giftedUnit.name}] as a gift!", locations, givingCityState.civName, giftedUnit.name)
+    }
+
+    fun turnsForGreatPersonFromCityState(): Int = (40 + -2 + Random().nextInt(5)) * gameInfo.gameParameters.gameSpeed.modifier.toInt()
+        // There seems to be some randomness in the amount of turns between receiving each great person,
+        // but I have no idea what the actual lower and upper bound are, so this is just an approximation
+    
```

Comment:
This can be found in civ5 files along with the entire function, but i forgot what the name of that thing was. Search through file contents and you should find it. 

 https://github.com/Gedemon/Civ5-DLL/blob/master/CvGameCoreDLL_Expansion1

I believe line 11540 in cvplayer.cpp file, but it is mentioned and modified in other files.